engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,609 @@
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.api.profiles.LoopConditionProfile;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.node.callable.ExecuteCallNode;
+import org.enso.interpreter.node.expression.builtin.meta.EqualsAnyNodeGen.InvokeEqualsNodeGen;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+import org.enso.interpreter.runtime.state.State;
+import org.enso.polyglot.MethodNames;
+
+@BuiltinMethod(
+    type = "Any",
+    name = "equals_builtin",
+    description = "Implementation of Any.=="
+)
+@GenerateUncached
+public abstract class EqualsAnyNode extends Node {
+
+  protected static String EQUALS_MEMBER_NAME = MethodNames.Function.EQUALS;
+
+  public static EqualsAnyNode build() {
+    return EqualsAnyNodeGen.create();
+  }
+  public abstract boolean execute(@AcceptsError Object self, @AcceptsError Object right);
+
+  /** Primitive values **/
+
+  @Specialization
+  boolean equalsLong(long self, long other) {
+    return self == other;
+  }
+
+  @Specialization
+  boolean equalsLongDouble(long self, double other) {
+    return (double) self == other;
+  }
+
+  @Specialization
+  boolean equalsDoubleLong(double self, long other) {
+    return self == (double) other;
+  }
+
+  @Specialization
+  boolean equalsIntLong(int self, long other) {
+    return (long) self == other;
+  }
+
+  @Specialization
+  boolean equalsLongInt(long self, int other) {
+    return self == (long) other;
+  }
+
+  @Specialization
+  boolean equalsBoolean(boolean self, boolean other) {
+    return self == other;
+  }
+
+  @Specialization
+  boolean equalsDouble(double self, double other) {
+    return self == other;
+  }
+
```

Comment:
If we have int==long and long==double, we should also have int==double to ensure that the relation is properly transitive.

Or is this handled somewhere else which I've missed?

Reply:
Also, do the `int`s show up anywhere in Enso? I was sure our numbers are always backed by a `long`.

If we are doing `int`, maybe we should do `short` too then?

Reply:
According to tests, everything seems fine for now. If there are any troubles in the future, we can just add a new specialization here - shouldn't be a problem. Resolving this conversation.

Reply:
Then either we do not get `int`s anywhere or the current implementation is invalid. I'm really opposed to this half-measure.

Let's either:
1. remove the specializations for `int` altogether 
2. OR add specializations for int-double pairs to make sure that the `==` relation on numbers is closed under transitivity (double==long==int).

If we go for (1) and tests are still passing, that sounds OK for me.
If tests weren't passing with (1), then we should ensure that if we go for (2), we add proper tests to cover such cases - probably need to create an `int` in Python and thread it through Enso to compare with a Decimal.

Reply:
`int==double` is handled in `equalsGeneric` fallback method. But I will add these specializations here, that won't hurt.

Removing any primitive specializations is not a good idea. Even though we don't get `int` from Enso itself, we may get `int` from some other language. In that case, we can at least avoid boxing by having these specializations.

TL;DR; I will add specialization for `int==double`, even though it is handled in `equalsGeneric`, therefore, the relation `==` between `int`, `long`, and `double`, is properly transitive.

Reply:
> Actually, I don't think equalsGeneric was handling this case correctly,

Good point, I have not realized that.